File: sdk/python/dogdata/resources/price.py
from __future__ import annotations

from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    from dogdata.client import DogData

EXCHANGES = ["kraken", "gateio", "mexc", "bitget", "pionex", "magiceden", "bitflow", "dogswap"]


class PriceResource:
    def __init__(self, client: DogData):
        self._client = client

    def current(self) -> dict[str, Any]:
        return self._client.request("/api/price/kraken")

    def from_exchange(self, exchange: str) -> dict[str, Any]:
        if exchange not in EXCHANGES:
            raise ValueError(f"Unknown exchange: {exchange}. Valid: {EXCHANGES}")
        return self._client.request(f"/api/price/{exchange}")

    def all(self) -> list[dict[str, Any]]:
        results = []
        for exchange in EXCHANGES:
            try:
                results.append(self.from_exchange(exchange))
            except Exception:
                pass
        return results

    def kraken(self) -> dict[str, Any]:
        return self.from_exchange("kraken")

    def gateio(self) -> dict[str, Any]:
        return self.from_exchange("gateio")

    def mexc(self) -> dict[str, Any]:
        return self.from_exchange("mexc")

    def bitget(self) -> dict[str, Any]:
        return self.from_exchange("bitget")

    def pionex(self) -> dict[str, Any]:
        return self.from_exchange("pionex")

    def magiceden(self) -> dict[str, Any]:
        return self.from_exchange("magiceden")

    def bitflow(self) -> dict[str, Any]:
        return self.from_exchange("bitflow")

    def dogswap(self) -> dict[str, Any]:
        return self.from_exchange("dogswap")

File: sdk/typescript/src/resources/price.ts
import type { DogData } from '../client.js';
import type { PriceData, AggregatedPrice } from '../types.js';

export type Exchange = 'kraken' | 'gateio' | 'mexc' | 'bitget' | 'bitflow' | 'dogswap';

export class PriceResource {
  constructor(private client: DogData) {}

  async current(): Promise<AggregatedPrice> {
    return this.client.request<AggregatedPrice>('/api/price/kraken');
  }

  async fromExchange(exchange: Exchange): Promise<PriceData> {
    return this.client.request<PriceData>(`/api/price/${exchange}`);
  }

  async all(): Promise<PriceData[]> {
    const exchanges: Exchange[] = ['kraken', 'gateio', 'mexc', 'bitget', 'bitflow', 'dogswap'];
    const results = await Promise.allSettled(
      exchanges.map(exchange => this.fromExchange(exchange))
    );
    return results
      .filter((r): r is PromiseFulfilledResult<PriceData> => r.status === 'fulfilled')
      .map(r => r.value);
  }

  async kraken(): Promise<PriceData> {
    return this.fromExchange('kraken');
  }

  async gateio(): Promise<PriceData> {
    return this.fromExchange('gateio');
  }

  async mexc(): Promise<PriceData> {
    return this.fromExchange('mexc');
  }

  async bitget(): Promise<PriceData> {
    return this.fromExchange('bitget');
  }

  async bitflow(): Promise<PriceData> {
    return this.fromExchange('bitflow');
  }

  async dogswap(): Promise<PriceData> {
    return this.fromExchange('dogswap');
  }
}
